Import custom attributes with GLTFDocumentExtension

huwpascoe opened the pull request:
## Proposal
Import ARRAY_CUSTOM# attributes with GLTFDocumentExtension using a function to specify layer name. Also added "Export Attributes" checkbox to the blender importer.
https://github.com/godotengine/godot-proposals/issues/10892

## Test Project
![image](https://github.com/user-attachments/assets/ae93dc78-5555-479f-9d81-8a2036a56099)
[custom_attrib_test.zip](https://github.com/user-attachments/files/23870955/custom_attrib_test.zip)

## changes
- Ability to combine 2 attributes per CUSTOM
- Created new GLTFAttributeMap resource to deal with the sheer number of possibilities
```gdscript
extends GLTFDocumentExtension

func _import_get_attribute_map(state: GLTFState, mesh_index: int) -> GLTFAttributeMap:
	var attrmap := GLTFAttributeMap.new()

	var mesh_name: String = state.json["meshes"][mesh_index]["name"]
	match mesh_name:
		"mesh_1": attrmap.color = "COLOR_1"
		"mesh_2": attrmap.color = "_MY_DATA"
		"custom":
			attrmap.custom0 = "_ROCK"
			attrmap.custom1 = "_MUD"
			attrmap.custom2 = ""
			attrmap.custom3 = "_GRASS"
			attrmap.custom3_mux = "_SNOW"

	return attrmap
```

## Supported attributes
- Mesh.ARRAY_TEX_UV (vec2)
- Mesh.ARRAY_TEX_UV2 (vec2)
- Mesh.ARRAY_COLOR (vec3, vec4)
- Mesh.ARRAY_CUSTOM# (scalar, vec2, vec3, vec4)

huwpascoe committed:
Import custom attributes

fire commented:
I'm having a hard time understanding the extension. Can you link or write a proposal in https://github.com/godotengine/godot-proposals?

Something like: I have a game workflow that uses blender I want the vertex attributes to import from blender, but it needs this pr. I can use the attributes for a shader effect.

aaronfranke commented on modules/gltf/gltf_document.cpp:
```diff
@@ -3136,6 +3136,64 @@ Error GLTFDocument::_parse_meshes(Ref<GLTFState> p_state) {
+							array[custom_i] = _decode_accessor_as_vec2(p_state, a[attrib_name], true, indices_mapping);
+							format = Mesh::ARRAY_CUSTOM_RG_FLOAT;
+						} break;
+						case GLTFAccessor::TYPE_VEC3: {
+							array[custom_i] = _decode_accessor_as_vec3(p_state, a[attrib_name], true, indices_mapping);
+							format = Mesh::ARRAY_CUSTOM_RGB_FLOAT;
+						} break;
+						case GLTFAccessor::TYPE_VEC4: {
+							array[custom_i] = _decode_accessor_as_floats(p_state, a[attrib_name], true, indices_vec4_mapping);
```
Why are Vec3s decoded as Vector3, but Vec4s are not decoded as Vector4?

huwpascoe replied:
There's not a decode_as_vec4 method, that's how it's done in the rest of the class. But come to think of it, they probably all need to be decoded as float, thanks for pointing that out.

aaronfranke commented on modules/gltf/extensions/gltf_document_extension.cpp:
```diff
@@ -34,6 +34,7 @@ void GLTFDocumentExtension::_bind_methods() {
 	// Import process.
 	GDVIRTUAL_BIND(_import_preflight, "state", "extensions");
 	GDVIRTUAL_BIND(_get_supported_extensions);
+	GDVIRTUAL_BIND(_remap_custom_attribute, "name");
```
This isn't documented, you need to run doctool.

huwpascoe commented:
https://github.com/godotengine/godot-proposals/issues/10892

huwpascoe commented:
Refactored and removed like a hundred duplicate lines of code.
```cpp
_decode_accessor_as_floats();
_decode_accessor_as_ints();
_decode_accessor_as_vec2();
_decode_accessor_as_vec3();
_decode_accessor_as_color();
_decode_accessor_as_quaternion();
_decode_accessor_as_xform2d();
_decode_accessor_as_basis();
_decode_accessor_as_xform();

// VVV

_decode_accessor<T>();
```

Multi-element scalars are now decoded with _decode_accessor<T, N>()
```cpp
array[Mesh::ARRAY_TANGENT] = _decode_accessor<float, 4>(p_state, a["TANGENT"], true, indices_mapping);
```

aaronfranke commented:
Since this PR alters the accessor code, it will have to wait until after https://github.com/godotengine/godot/pull/94165 which also touches this code.

Flynsarmy commented:
> Since this PR alters the accessor code, it will have to wait until after #94165 which also touches this code.

PR in question was just merged so we should be good to continue here.

fire commented:
@huwpascoe Many people think this is a good enhancement. Would you happen to have time to rebase it? <3

huwpascoe commented:
@fire

Would it be better to keep as an extension or promote it to import options?
![image](https://github.com/user-attachments/assets/52dc1357-7b32-49c2-a2ce-216343f69d60)

There are trade-offs for each.

huwpascoe commented:
> Remember to take this PR out of draft if it's ready :)

Don't worry! I haven't forgotten.

> I am trending towards promoting it to import options

Right, I've determined that the UI approach is just too much of a mess for what is a technical feature. So I'll be moving forward with the original plan to add the method to DocumentExtension.

huwpascoe commented:
I'll make a test project soon.

aaronfranke commented on modules/gltf/gltf_document.cpp:
```diff
@@ -1571,68 +1594,109 @@ Error GLTFDocument::_parse_meshes(Ref<GLTFState> p_state) {
-						cur_custom.write[uv_i * num_channels + 0] = texcoord_first[uv_i].x;
-						cur_custom.write[uv_i * num_channels + 1] = texcoord_first[uv_i].y;
+			if (using_custom_attributes) {
+				for (int custom_i = 0; custom_i < 4; ++custom_i) {
```
Godot's style is to use postfix `++`.

```suggestion
				for (int custom_i = 0; custom_i < 4; custom_i++) {
```

lyuma reviewed (commented):
I have some comments about the design that I'd like a chance to discuss.

Firstly, I have concerns about the usability of depending on a customized GLTFDocumentExtension to perform the remapping... it certainly could work, but document extensions are registered globally (project-wide), so any attempt to use this feature would have to condition the remapping on some condition or add a setting to the import dock. If I were to copy the sample code as written, it would break all other glTF models in my project, even if they do not have that custom attribute. But explaining all this is probably too much for a method description, so I'm not sure how to clarify this.

Second, as I noticed, if I use this to override any custom attributes, then I can no longer combine two UV channels into one custom channel which limits the number of custom 2-channel UVs I can import. I feel like maybe something simple like "TEXCOORD_1,TEXCOORD_2" could work to handle that case. I think it only needs to support `2 x vec2` or `1 x anything` cases , so no need to handle every combination.

Basically the usecase is for compacting multiple 2-channel UVs into a CUSTOM, and this is Godot's current behavior so it would be nice to support it for that reason without needing a whole special case. At a minimum, it should be possible to return "" for an individual custom channel to get the current behavior without losing all the other custom channels.

huwpascoe commented:
Individual attribute querying is gone, now extensions only need to return a GLTFAttributeMap resource. With defaults set to how texture UVs were originally handled.

Any 2 attributes can combine into a CUSTOM provided they fit within a vec4. For example, vec3+scalar is valid.

<img width="339" height="358" alt="image" src="https://github.com/user-attachments/assets/570f115c-ad8d-4d8b-9c5d-d38173ccbb0d" />

Calinou commented on modules/gltf/doc_classes/GLTFAttributeMap.xml:
```diff
@@ -0,0 +1,62 @@
+		[b]Note:[/b] It may be necessary to disable [code]generate_lods[/code] in the mesh import settings to preserve custom data.
+
+		Multiplexing is supported with CUSTOM* attributes. Setting custom* and custom*_mux will combine them into the one array.
+		The attributes must total no larger than vec4. If the _mux attribute is missing, it will act like only the first was set.
```
```suggestion
		Controls how glTF attributes are mapped to the resulting [Mesh]. Used by [method GLTFDocumentExtension._import_get_attribute_map].
		Multiplexing is supported with [code]CUSTOM*[/code] attributes. Setting [code]custom*[/code] and [code]custom*_mux[/code] will combine them into the one array.
		A pair of [code]custom*[/code] and [code]custom*_mux[/code] attributes must total no larger than [code]vec4[/code]. For instance, you can set [member custom0] to a scalar and [member custom0_mux] to a [code]vec3[/code], but you cannot set [member custom0] to a [code]vec2[/code] and [member custom0_mux] to a [code]vec3[/code] (as this would require 5 values).
		If the [code]custom*_mux[/code] attribute that is requested is missing, it will act like only the first was set.
		[b]Note:[/b] It may be necessary to disable [code]generate_lods[/code] in the mesh import settings to preserve custom data.
```

Calinou commented on modules/gltf/doc_classes/GLTFAttributeMap.xml:
```diff
@@ -0,0 +1,62 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="GLTFAttributeMap" inherits="Resource" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../doc/class.xsd">
+	<brief_description>
+		Controls how GLTF attributes are mapped.
```
```suggestion
		Controls how glTF attributes are mapped on import.
```

lyuma commented:
Some thoughts from the asset meeting.

I can see that this feature could be useful for advanced users. We need to balance this with the additional complexity, and the lack of discoverability of this feature.

Since #96748 was approved, this change will have to be rebased on top of it. In particular, it may impact the order of the UV attributes based on `_gltf_primary_texture_coord` and `_gltf_secondary_texture_coord` meta in each of the Material3D attached to the various mesh primitives, often leading TEXCOORD_0 and TEXCOORD_1 to be swapped.

So we can automatically do this fix on the default GLTFAttributeMap. We need to consider how this change may impact extensions that set a custom one. 

Finally, my other concern is this feature is difficult to discover and therefore has a very limited set of users who will know about or use this. Ideally, it would be visible in the advanced importer for example something you can set on each mesh/material to override the GLTFAttributeMap resource slot, and then of course can be again overridden by extensions.

Christakou commented:
Just wondering if there's been progress on this as I just encountered a usecase in my engine usage where this would be super useful
